src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -0,0 +1,89 @@
+package com.puppetlabs.puppetserver.pool;
+
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import com.puppetlabs.puppetserver.pool.LockablePool;
+
+public final class JRubyPool<E> implements LockablePool<E> {
+    private final LinkedBlockingDeque<E> live_queue;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
+    private final Set<E> registeredElements = new CopyOnWriteArraySet<>();
+
+    public JRubyPool(int size) {
+        live_queue = new LinkedBlockingDeque<>(size);
+    }
+
+    @Override
+    public void register(E e) throws InterruptedException {
+        live_queue.putLast(e);
+        registeredElements.add(e);
+    }
+
+    @Override
+    public E borrowItem() throws InterruptedException, IllegalStateException {
+        if (lock.isWriteLockedByCurrentThread()) {
+          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+        }
+        E item = live_queue.takeFirst();
+        lock.readLock().lock();
+        return item;
+    }
+
+    @Override
+    public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws InterruptedException, IllegalStateException {
+        if (lock.isWriteLockedByCurrentThread()) {
+          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+        }
+        E item = live_queue.pollFirst(timeout, unit);
+        lock.readLock().lock();
+        return item;
+    }
```

Comment:
It'd be nice if we could refactor the two methods above to share some common code, but, in Java 7 there's no way to do that without a bunch of extra interface junk that would almost certainly end up being less maintainable than the duplication you've got here now, so it's probably better to leave it like this.